Reorder and expand porting guidance steps

twsouthwick opened the pull request:
## Summary

The steps as listed work, but assume some knowledge and aren't in the
best order. This change updates the list to be what we have seen work
the most often with people doing changes in live codebases.

Fixes #16182

twsouthwick committed:
Reorder and expand porting guidance steps

The steps as listed work, but assume some knowledge and aren't in the
best order. This change updates the list to be what we have seen work
the most often with people doing changes in live codebases.

twsouthwick committed:
Update some text

twsouthwick commented:
Looks like there's a build error unrelated to my change.

tdykstra commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,68 @@ You might have code that currently runs on the .NET Framework that you're intere
 ## Overview of the porting process
 
-We recommend you use the following process when porting your project to .NET Core:
+We recommend you use the following process when porting your project to .NET Core. Each of these steps introduce potential places for behavior changes, so ensure that you adequately test your library or application before continuing on to later steps. The first steps are to get your project ready for a switch to .NET Standard/.NET Core. If you have unit tests, it is best to convert them first so  ...
```
```suggestion
We recommend you use the following process when porting your project to .NET Core. Each of these steps introduces potential places for behavior changes, so ensure that you adequately test your library or application before continuing on to later steps. The first steps are to get your project ready for a switch to .NET Standard or .NET Core. If you have unit tests, it's best to convert them first so that you can continue testing changes in the product you're working on. Because porting to .NET Core is such a significant change to your codebase, it's highly recommended to port your test projects so that you can run tests as you port your code over. MSTest, xUnit, and NUnit all work on .NET Core.
```

twsouthwick replied:
Done

gewarren commented:
Close/reopen to trigger a new build.

tdykstra commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,68 @@ You might have code that currently runs on the .NET Framework that you're intere
+
+   .NET Core uses a simplified (and different) [project file format](../tools/csproj.md) than .NET Framework. You'll need to convert your project files into this format to continue. This project style allows you to also target .NET Framework, which at this point you'll still want to target.
+
+   You can attempt to port smaller solutions or individual projects in one operation to the .NET Core project file format with the [dotnet try-convert](https://github.com/dotnet/try-convert) tool. `dotnet try-convert` is not guaranteed to work for all your projects, and it may cause subtle changes in behavior that you depended on. Use it as a _starting point_ that automates the basic things that  ...
```
```suggestion
   You can attempt to port smaller solutions or individual projects in one operation to the .NET Core project file format with the [dotnet try-convert](https://github.com/dotnet/try-convert) tool. `dotnet try-convert` is not guaranteed to work for all your projects, and it may cause subtle changes in behavior that you depended on. Use it as a _starting point_ that automates the basic things that can be automated. It isn't a guaranteed solution to migrating a project, as there are many differences in the targets used by the SDK style projects compared to the old-style project files.
```

twsouthwick replied:
Done

tdykstra commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,68 @@ You might have code that currently runs on the .NET Framework that you're intere
    This step ensures that you can use API alternatives for .NET Framework-specific targets when .NET Core doesn't support a particular API.
 
-2. Use the [.NET Portability Analyzer](../../standard/analyzers/portability-analyzer.md) to analyze your assemblies and see if they're portable to .NET Core.
+1. Update all dependencies to the latest version. Projects may be using older versions of libraries that may not have .NET Standard support. However, later versions may support it with a simple switch. This may require code changes depending on any breaking changes in libraries.
```
```suggestion
1. Update all dependencies to the latest version. Projects may be using older versions of libraries that may not have .NET Standard support. However, later versions may support it with a simple switch. This may require code changes if there are breaking changes in libraries.
```

twsouthwick replied:
Done

tdykstra commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,68 @@ You might have code that currently runs on the .NET Framework that you're intere
-3. Install the [.NET API analyzer](../../standard/analyzers/api-analyzer.md) into your projects to identify APIs that throw <xref:System.PlatformNotSupportedException> on some platforms and some other potential compatibility issues.
+   The API Portability Analyzer tool analyzes your compiled assemblies and generates a report. This report shows a high-level portability summary and a breakdown of each API you're using that isn't available on NET Core. While using the tool, only submit the individual project you are converting to focus on the API changes that are potentially needed. Many of the APIs have equivalent availability ...
+
+   There are some cases where APIs are not equivalent and you'll need to do some compiler preprocessor directives (i.e. `#if NET45`) to special case the platforms. At this point, you're project will still be targeting .NET Framework. For each of these targeted cases, it is recommended to use well-known conditionals that can be understood as a scenario.  For example, AppDomain support in .NET Core ...
```
```suggestion
   There are some cases where APIs are not equivalent and you'll need to do some compiler preprocessor directives (i.e. `#if NET45`) to special case the platforms. At this point, your project will still be targeting .NET Framework. For each of these targeted cases, we recommend that you use well-known conditionals that can be understood as a scenario.  For example, AppDomain support in .NET Core is limited, but for the scenario of loading and unloading assemblies, there is a new API that is not available in .NET Core. A common way to handle this in code would be something like this:
```

twsouthwick replied:
Done

tdykstra commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,68 @@ You might have code that currently runs on the .NET Framework that you're intere
    This tool is similar to the portability analyzer, but instead of analyzing if code can build on .NET Core, it analyzes whether you're using an API in a way that will throw a <xref:System.PlatformNotSupportedException> at run time. Although this isn't common if you're moving from .NET Framework 4.7.2 or higher, it's good to check. For more information about APIs that throw exceptions on .NET Co ...
 
-4. Convert all of your `packages.config` dependencies to the [PackageReference](/nuget/consume-packages/package-references-in-project-files) format with the [conversion tool in Visual Studio](/nuget/consume-packages/migrate-packages-config-to-package-reference).
+1. If you are using Windows-specific APIs (such as registry access), you can install the [Windows Compatibility Pack](./windows-compat-pack.md).
```
```suggestion
1. If you're using Windows-specific APIs (such as registry access), you can install the [Windows Compatibility Pack](./windows-compat-pack.md).
```

twsouthwick replied:
Done

tdykstra reviewed (approved):
LGTM Some minor typo and wording suggestions.

gewarren commented:
I rebuilt master. Closing/reopening again. If this doesn't work, I'll file a sitehelp issue.

twsouthwick commented:
Looks like there's still a `BackendEngineExceptionError`

tdykstra commented:
I ran into BackendEngineExceptionError on another PR a couple days ago and reported it, and the recommendation was to have the person whose fork was causing the problem to log on to the OPS portal (ops.microsoft.com). That was said to provide OPS with an auth token it needs.  It's a bug they're working on fixing, if this is the same error.

twsouthwick committed:
Add solution wide

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
 
+Porting to .NET Core (or .NET Standard) from .NET Framework can become complicated very quickly depending on the complexity of the project. This document should help with identifying the main strategies that have been employed by users to successfully convert their code bases. This will address the conversion at two levels: solution-wide and project specific.
+
+Each of these steps introduce potential places for behavior changes, so ensure that you adequately test your library or application before continuing on to later steps. The first steps are to get your project ready for a switch to .NET Standard/.NET Core. If you have unit tests, it is best to convert them first so that you can continue testing changes in the product you are working on. Because po ...
```
I must say, this sounds a bit scary :) Depending on the scenario, migration *can* be a lot of work. But other scenarios (libraries and desktop apps, specifically) often port over without too much pain. We may want to rephrase a bit to not be too intimidating.

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
+
+When there are multiple projects in a solution the porting can seem more complicated because you must address projects in a specific order. The conversion process should be a bottom-up approach, where the projects with no dependencies on other projects in the solution are converted first, and continue up through the whole solution.
+
+In order to identify the order projects should be identified, you can use the following tools:
```
Type: replace 'identified' with 'migration' (or some similar word).

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
+
+   You can attempt to port smaller solutions or individual projects in one operation to the .NET Core project file format with the [dotnet try-convert](https://github.com/dotnet/try-convert) tool. `dotnet try-convert` is not guaranteed to work for all your projects, and it may cause subtle changes in behavior that you depended on. Use it as a _starting point_ that automates the basic things that  ...
+
+1. Convert all of your `packages.config` dependencies to the [PackageReference](/nuget/consume-packages/package-references-in-project-files) format with the [conversion tool in Visual Studio](/nuget/consume-packages/migrate-packages-config-to-package-reference).
```
I would put this before switching csproj formats, as I always do it first. Do SDK-style projects even work with packages.config?

twsouthwick replied:
Good catch. Yeah, they need to be switched

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
 
-3. Install the [.NET API analyzer](../../standard/analyzers/api-analyzer.md) into your projects to identify APIs that throw <xref:System.PlatformNotSupportedException> on some platforms and some other potential compatibility issues.
+   The API Portability Analyzer tool analyzes your compiled assemblies and generates a report. This report shows a high-level portability summary and a breakdown of each API you're using that isn't available on NET Core. While using the tool, only submit the individual project you are converting to focus on the API changes that are potentially needed. Many of the APIs have equivalent availability ...
+
```
It may be worth spending a couple sentences here talking about how to interpret an APIPort report since that's an area we see a lot of confusion. I'd stress that the single % number that's produced is not that important. More important is the list of specific APIs that are used and not available on NetCore/NetStandard.

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
-4. Convert all of your `packages.config` dependencies to the [PackageReference](/nuget/consume-packages/package-references-in-project-files) format with the [conversion tool in Visual Studio](/nuget/consume-packages/migrate-packages-config-to-package-reference).
+1. If you are using Windows-specific APIs (such as registry access), you can install the [Windows Compatibility Pack](./windows-compat-pack.md).
+
+1. At this point, you can switch to targeting .NET Core (generally for applications) or .NET Standard (for libraries).
```
Should we discuss in any more detail how a user decides whether to target NetCore or NetStandard? It may be out of scope for this doc, but I could also imagine a sentence or two explaining that apps need to use NetCore whereas it's a best practice to target libraries at NetStandard for portability unless there are dependencies (recent ASP.NET Core or EF Core libraries, for example) that require NetCore.

twsouthwick replied:
I've added some text for that, but couldn't find a link. Do you know of docs that address that choice?

mjrousos replied:
Now that you mention it, I don't know of any docs on that topic. May need to just give a quick description here for the moment. In the future, that may be something we want to add somewhere in the .NET docs.

twsouthwick replied:
I've added some text, and agree we should look into a more detailed page about it

mjrousos commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,91 @@ You might have code that currently runs on the .NET Framework that you're intere
 
 >[!div class="nextstepaction"]
 >[Analyze dependencies](third-party-deps.md)
+>[Package NuGet Package](../deploying/creating-nuget-packages.md)
```
I realize that ASP.NET -> ASP.NET Core is only one specific type of migration, but I think it might be valuable to have links to our ASP.NET Core migration docs here because that transition is both common enough and more-complicated-than-other-migrations enough that a lot of people seeing this doc should probably also see those ones.

twsouthwick replied:
Added

mjrousos reviewed (approved):
Looks good. I added a few comments/suggestions.

twsouthwick committed:
Updates per feedback

twsouthwick committed:
Add newline

twsouthwick commented:
@tdykstra I've addressed your feedback and @mjrousos. I've included some links such as `https://docs.microsoft.com/en-us/visualstudio/` and wasn't sure if that's the appropriate way to do it (I see lots of relative links, but couldn't figure out a relative for visualstudio or aspnetcore as those docs don't look to be in this repo.

twsouthwick committed:
Clean up links per warnings

twsouthwick committed:
Clean up trailing spaces per linter

twsouthwick committed:
Remove extra word

twsouthwick commented:
@tdykstra @gewarren any comment on the question about links for docs outside the repo? I'll assume if I don't hear back I can merge this in.

tdykstra commented:
@twsouthwick You can use a relative link starting with `/` for any URL that begins `docs.microsoft.com` regardless of which repo it comes from. The equivalent of `https://docs.microsoft.com/visualstudio/` is `/visualstudio/`. These links will not work in the staging site but will in production.

twsouthwick committed:
UPdate links

cartermp commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,96 @@ You might have code that currently runs on the .NET Framework that you're intere
+
+In order to identify the order projects should be migrated, you can use the following tools:
+
+- [Dependency Diagrams in Visual Studio](/visualstudio/modeling/create-layer-diagrams-from-your-code) can create a directed graph of the code in a solution.
```
Does the feature work for C++/CLI and F# projects? Which VS component or workload is also required to install these? I would suspect the majority of customers don't have the tooling installed unless it's a default in the most popular workloads.

cartermp commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,96 @@ You might have code that currently runs on the .NET Framework that you're intere
+
 We recommend you use the following process when porting your project to .NET Core:
 
+1. Convert all of your `packages.config` dependencies to the [PackageReference](/nuget/consume-packages/package-references-in-project-files) format with the [conversion tool in Visual Studio](/nuget/consume-packages/migrate-packages-config-to-package-reference).
```
Note that this only works for old-style, unflavored C# and VB projects. ASP.NET, F#, and C++/CLI don't support this (C++/CLI doesn't support PackageReference at all yet, but it is planned for 16.8). Same goes for Xamarin, IIRC - which will be relevant for .NET 5.

cartermp commented on docs/core/porting/index.md:
```diff
@@ -13,35 +13,96 @@ You might have code that currently runs on the .NET Framework that you're intere
-3. Install the [.NET API analyzer](../../standard/analyzers/api-analyzer.md) into your projects to identify APIs that throw <xref:System.PlatformNotSupportedException> on some platforms and some other potential compatibility issues.
+   While reading the reports generated by the analyzer, the important information is the actual APIs that are being used and not necessarily the percentage of support for the target platform. Many APIs have equivalent options in .NET Standard/Core, and so understanding the scenarios your library or application needs the API for will help determine the implication for portability.
+
+   There are some cases where APIs are not equivalent and you'll need to do some compiler preprocessor directives (i.e. `#if NET45`) to special case the platforms. At this point, you're project will still be targeting .NET Framework. For each of these targeted cases, it is recommended to use well-known conditionals that can be understood as a scenario.  For example, AppDomain support in .NET Core ...
```
> there is a new API that is not available in .NET Core

Should this say "there is a new API that is not available on .NET Framework"? I assume you're referring to `AssemblyLoadContext`? Interestingly enough, the NuGet Package claims support for [.NET Standard](https://www.nuget.org/packages/System.Runtime.Loader/), so this might be inviting a little bit of trouble as an example since NS1.5 _claims_ to be compatible with .NET Framework.

cartermp reviewed (commented):
Sorry for the late review here, didn't see the PR. Overall I think the changes are great. Some quick nits/questions.
